### ISSUE TITLE
Issue with pure gan loss i.e.( L1 weight = 0 Gan weight = 1)

### ISSUE THREAD

**Issue by nikhilchh** [affinelayer/pix2pix-tensorflow#85]
I wanted to see results for pure gan loss for image inpainting task.
Tried Image inpainting for faces. Pure Gan loss give bad results even after 2 days of training. It's just draws random noise, no sign of faces. 
But works fine if I use l1 weight = 100+ and Gan weight = 1

**Comment by julien2512** [comment]
Yes.
L1+GAN > L1 > GAN

**Comment by nikhilchh** [comment]
Yes but pure gan should atleast give something close to face and not just noisy output.

**Comment by julien2512** [comment]
I am checking it on my volcanoes.

**Comment by julien2512** [comment]
Here is the last steps :
> progress  epoch 44  step 258  image/sec 0.2  remaining 4468m
> discrim_loss 0.784925
> gen_loss_GAN 1.53159
> gen_loss_L1 0.536851
> progress  epoch 44  step 308  image/sec 0.2  remaining 4464m
> discrim_loss 0.925797
> gen_loss_GAN 1.42804
> gen_loss_L1 0.507922

We can see that discrim_loss still won.
gen_loss_L1, although no gradient apply on it, seems to have increase !
That mean the image is worst than before.

Let is be a few more.

**Comment by julien2512** [comment]
Now here is the L1 loss :
![generator_loss_l1](https://user-images.githubusercontent.com/838473/33803665-2561f2d4-dd95-11e7-991f-3e92a80b565d.png)
after big increase, it slowly slowly very slowly decreases (remember there is no gradient apply to it. It is a measure of the learning quality here).

The discriminator & generator start to fight against each other :
![discriminator_loss](https://user-images.githubusercontent.com/838473/33803675-4d7f8fa6-dd95-11e7-866c-01442bb78c01.png)
![generator_loss_l1](https://user-images.githubusercontent.com/838473/33803676-5048a880-dd95-11e7-8faa-b6efc2f5ffe5.png)

**Comment by julien2512** [comment]
A last snapshot before I ended it :

> progress  epoch 140  step 84  image/sec 0.2  remaining 1737m
> discrim_loss 0.816318
> gen_loss_GAN 1.66294
> gen_loss_L1 0.442908
> recording summary
> progress  epoch 140  step 134  image/sec 0.2  remaining 1733m
> discrim_loss 0.868367
> gen_loss_GAN 1.56294
> gen_loss_L1 0.428938
> progress  epoch 140  step 184  image/sec 0.2  remaining 1730m
> discrim_loss 0.875951
> gen_loss_GAN 1.60302
> gen_loss_L1 0.45576
> recording summary
> progress  epoch 140  step 234  image/sec 0.2  remaining 1726m
> discrim_loss 0.860077
> gen_loss_GAN 1.55999
> gen_loss_L1 0.459956
> saving model
> progress  epoch 140  step 284  image/sec 0.2  remaining 1723m
> discrim_loss 0.823654
> gen_loss_GAN 1.63946
> gen_loss_L1 0.490804

It is very slow, but pure GAN on hesse's tensorflow is working with my volcanoes pics.
I will use L1+GAN for my own purposes definitly !

**Comment by myway0101** [comment]
@nikhilchh @julien2512 

Hi guys,

As I understand, to use pure gan, we can set L1 weight = 0.
How to use the only L1 loss in pix2pix.py? 
Just set  a.gan_weight = 0 ? But in this case, i think discrim_loss is still updating.

Many thanks!

**Comment by julien2512** [comment]
Yes, you got it, as the generator is used here :
> with tf.name_scope("fake_discriminator"):
>         with tf.variable_scope("discriminator", reuse=True):
>             # 2x [batch, height, width, channels] => [batch, 30, 30, 1]
>             predict_fake = create_discriminator(inputs, outputs)
because :
> with tf.variable_scope("generator") as scope:
>         out_channels = int(targets.get_shape()[-1])
>         outputs = create_generator(inputs, out_channels)

You will need to remove :
>    with tf.name_scope("discriminator_train"):
>         discrim_tvars = [var for var in tf.trainable_variables() if var.name.startswith("discriminator")]
>         discrim_optim = tf.train.AdamOptimizer(a.lr, a.beta1)
>         discrim_grads_and_vars = discrim_optim.compute_gradients(discrim_loss, var_list=discrim_tvars)
>         discrim_train = discrim_optim.apply_gradients(discrim_grads_and_vars)

or to propose a parameter (think twice about it).

**Comment by myway0101** [comment]
@julien2512
Thank you. 
So only use of l1 loss means that only use of generator? Not discriminator?

**Comment by julien2512** [comment]
Yes. The L1 loss is only : 
> gen_loss_L1 = tf.reduce_mean(tf.abs(targets - outputs))

the discriminator is related to GAN (discriminator vs generator !).
